### PR TITLE
Flax Speech-Encoder-Decoder Model

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Currently, Speech-Encoder-Decoder Models are only available in PyTorch. This PR adds the **Flax** Speech-Encoder-Decoder Model and all accompanying tests.

The accuracy of the Flax Speech-Encoder-Decoder Model is validated through cross-test comparison with the corresponding PyTorch Speech-Encoder-Decoder Model. The final logit values agree to within a `4e-2` threshold, making certain that the Flax Model operates as intended.

### CODE REVIEW

**Review by patrickvonplaten — COMMENTED** [review]
Looks pretty cool already! I think in a next step it would be good to add a testing file as well (similar to https://github.com/huggingface/transformers/blob/master/tests/test_modeling_speech_encoder_decoder.py but we only need the Wav2Vec2Bert ... testing class for now)

**Comment by patrickvonplaten** [comment]
> The issue being encountered occurs when the model weights are initialized: the `encoder_hidden_states` don’t match the shape of the `decoder_hidden_states`, meaning that the cross-attention layers throw an error when the decoder is initialised. I have been inspecting the code to try and establish where the two deviate in shape. For reference, running this code produces the error:
> 
> ```python
> from transformers import FlaxSpeechEncoderDecoderModel
> # Initialize a wav2vec2-2-gpt2 from pretrained wav2vec2 and gpt2 models
> # Note that the cross-attention layers will be randomly initialized
> model_fx = FlaxSpeechEncoderDecoderModel.from_encoder_decoder_pretrained("patrickvonplaten/dummy_wav2vec2_with_adapter", "gpt2", encoder_from_pt=True)
> ```

This already looks like a pretty edge-case loading. Could we maybe in a first step try to see whether the following works:

```py
from transformers import FlaxSpeechEncoderDecoderModel
# Initialize a wav2vec2-2-gpt2 from pretrained wav2vec2 and gpt2 models
# Note that the cross-attention layers will be randomly initialized
model_fx = FlaxSpeechEncoderDecoderModel.from_encoder_decoder_pretrained("patrickvonplaten/dummy_wav2vec2_with_adapter", "gpt2")
```
? 

Meaning that the weights don't have to be converted from PT to Flax. For this you probably first need to do:

```py
from transformers import FlaxWav2Vec2Model

model = FlaxWav2Vec2Model.from_pretrained("patrickvonplaten/dummy_wav2vec2_with_adapter", from_pt=True)
model.push_to_hub("<your-url>")
```

**Comment by patrickvonplaten** [comment]
@sanchit-gandhi,

Just pushed a fix for the problem. The reason was that the encoder downsamples the seq_len while the decoder does not. This means that we can't have the same encoder and decoder input lengths. 

The script still fails, but now because the attention_mask is not correct. Could you take a look here? We should implement `_get_feature_vector_attention_mask` now as you correctly noted :-)

**Comment by sanchit-gandhi** [comment]
Quick question about our PT-Flax cross-test thresholds. I noticed that in the `test_modeling_flax_common.py`, we use a threshold of 4e-2 when comparing Flax and PT outputs (https://github.com/huggingface/transformers/blob/60ba48205e4ec070780e8cb8d461421b77432bad/tests/test_modeling_flax_common.py#L192), whereas in `test_modeling_flax_encoder_decoder.py` the threshold is set much more aggressively at 1e-5 (https://github.com/huggingface/transformers/blob/60ba48205e4ec070780e8cb8d461421b77432bad/tests/test_modeling_flax_encoder_decoder.py#L261). Is there a reason why the threshold is set so much lower in the latter?

**Comment by patrickvonplaten** [comment]
So if we see differences of like 0.05 we are still pretty sure that the model is correctly implemented in Flax

File: src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+
+        ```python
+        >>> from transformers import FlaxSpeechEncoderDecoderModel, BartTokenizer
+
+        >>> # load a fine-tuned wav2vec2-2-bart model
+        >>> model = FlaxSpeechEncoderDecoderModel.from_pretrained("patrickvonplaten/wav2vec2-2-bart-large")
+        >>> # load output tokenizer
+        >>> tokenizer_output = BartTokenizer.from_pretrained("facebook/bart-large")
+
+        >>> inputs = jnp.ones((2, 5000), dtype=jnp.float32)
```
Comment (patrickvonplaten) [inline src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py:663]: Examples are usually the first thing people read so it would be nice to have a sensible example here. Could you modify the example a bit according to: https://github.com/huggingface/transformers/blob/935a76d90d43be58b62efce28e713c59d21b2191/src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py#L468

where as we do use the `patrickvonplaten/wav2vec2-2-bart-large` model though (can you upload it in Flax to this repo?). The output should give some sensible results here

File: src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+                - To update the encoder configuration, use the prefix *encoder_* for each configuration parameter.
+                - To update the decoder configuration, use the prefix *decoder_* for each configuration parameter.
+                - To update the parent model configuration, do not use a prefix for each configuration parameter.
+
+                Behaves differently depending on whether a `config` is provided or automatically loaded.
+
+        Example:
+
+        ```python
+        >>> from transformers import FlaxSpeechEncoderDecoderModel
```
Comment (patrickvonplaten) [inline src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py:796]: nice example

File: tests/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+            "config": config,
+            "inputs": inputs,
+            "attention_mask": attention_mask,
+            "decoder_config": decoder_config,
+            "decoder_input_ids": decoder_input_ids,
+            "decoder_attention_mask": decoder_attention_mask,
+            "encoder_hidden_states": encoder_hidden_states,
+        }
+
+    @slow
```
Comment (patrickvonplaten) [inline tests/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py:492]:
```suggestion
    @slow
    @require_torch
```

Comment (patrickvonplaten) [reply tests/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py:492]: Ok for now. We should clean this up in a follow-up PR though and add a slow Flax test here which does not rely on PyTorch. The easiest is probably to add a test that verifies that https://huggingface.co/patrickvonplaten/wav2vec2-2-bart-large gives sensible outputs. Think we'll need some code changes to FlaxBart for this though, so maybe let's do it in a follow-up PR.

File: src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/speech_encoder_decoder/modeling_flax_speech_encoder_decoder.py:663]: Actually, you'll need a FlaxBartForDecoder model for this. So let's maybe just do it in a follow-up PR

**Review by patrickvonplaten — APPROVED** [review]
Ok to merge for me for now. Would be nice to update the docstring example and the slow test in a follow-up PR

File: tests/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py (hunk 3 above)
Comment (sanchit-gandhi) [reply tests/speech_encoder_decoder/test_modeling_flax_speech_encoder_decoder.py:492]: Yes, we first need to incorporate a Causal LM to FlaxBart before being able to verify this. I'll target address this in a future PR.
